Make several TableOpsImpl methods atomic

DomGarguilo opened the pull request:
Fixes #5494 
This PR replaces places were multiple calls to set or remove properties were being done with a single modify properties block to make things atomic. The affected methods are:
* setSamplerConfiguration
* clearSamplerConfiguration
* setLocalityGroups

DomGarguilo committed:
Make setLocalityGroups() atomic

DomGarguilo committed:
make samplerConfig methods atomic

dlmarion commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1220,42 +1220,33 @@ public Map<String,String> getTableProperties(final String tableName)
-    for (Entry<String,Set<Text>> entry : groups.entrySet()) {
-      Set<Text> colFams = entry.getValue();
-      String value = LocalityGroupUtil.encodeColumnFamilies(colFams);
-      setPropertyNoChecks(tableName, Property.TABLE_LOCALITY_GROUP_PREFIX + entry.getKey(), value);
```
This is not referencing the `.getKey()` method of the property. Is this a bug? If so, does it exist in earlier versions that need to be fixed (2.1, 3.0)?

DomGarguilo replied:
Looks like its not a bug. `Property.toString()` is as follows:

```java
  @Override
  public String toString() {
    return this.key;
  }
```

Confirmed in an `uno jshell` shell too:
```
jshell> import org.apache.accumulo.core.conf.Property

jshell> String test = Property.TABLE_LOCALITY_GROUP_PREFIX + "";
test ==> "table.group."

jshell> String test2 = Property.TABLE_LOCALITY_GROUP_PREFIX.getKey() + "";
test2 ==> "table.group."
```

Definitely seems better to explicitly use `getKey()` but functionally the same.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1220,42 +1220,33 @@ public Map<String,String> getTableProperties(final String tableName)
-        if (!groups.containsKey(group)) {
-          removePropertyNoChecks(tableName, property);
+      // remove any stale locality groups that were previously set
+      for (String key : new ArrayList<>(properties.keySet())) {
```
Is there a reason to convert `Set<String>` to `ArrayList<String>`?

DomGarguilo replied:
No reason for ArrayList specifically, just needed to create a new collection to iterate over so we are not iterating over and removing from the props map at the same time.

DomGarguilo replied:
Actually took another look and ended up using `removeIf()` here which avoids the need to create a new collection and generally cleans the code up more.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1862,15 +1853,14 @@ private void doTableFateOperation(String tableOrNamespaceName,
     String prefix = Property.TABLE_SAMPLER_OPTS.getKey();
-    for (Entry<String,String> entry : getProperties(tableName)) {
-      String property = entry.getKey();
+    for (String property : new ArrayList<>(props.keySet())) {
```
Is there a reason to convert `Set<String>` to `ArrayList<String>`?

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into makeTableOperationsImplAtomic

DomGarguilo committed:
simplfy using removeIf()

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into makeTableOperationsImplAtomic

kevinrr888 reviewed (approved):
LGTM, code is also quite a bit more succinct now

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1862,37 +1852,31 @@ private void doTableFateOperation(String tableOrNamespaceName,
   @Override
   public void clearSamplerConfiguration(String tableName)
-      throws AccumuloException, TableNotFoundException, AccumuloSecurityException {
+      throws AccumuloException, AccumuloSecurityException {
```
Why would this no longer throw TableNotFoundException? That seems odd.

DomGarguilo replied:
Nothing in this method throws `TableNotFoundException` anymore I guess. In the old code, `clearSamplerOptions()` threw it when it called `getProperties()` but all of that has been replaced with the `modifyProperties()` block which does not throw `TableNotFoundException`.

ctubbsii replied:
I would expect modifyProperties to throw TableNotFoundException. At least, some error should happen if the table doesn't exist... I'm not sure what error that is. But, I think this PR should wait until we figure that out, because this changes the behavior, and we shouldn't change the API behavior for the case when the table doesn't exist.

DomGarguilo replied:
So it turns out some of the TableOps methods wrap TableNotFoundException in AccumuloException. Created #5595 to add a bit of clarity for those methods. I am not sure why some methods do this. It seems like it would be better to just throw it directly (might avoid confusion like this scenario).

Anyways, I dont think this is a blocker for this PR.

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into makeTableOperationsImplAtomic

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into makeTableOperationsImplAtomic

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into makeTableOperationsImplAtomic

DomGarguilo committed:
Unwrap TNFE for changed methods

DomGarguilo commented:
> Why would this no longer throw TableNotFoundException? That seems odd.

@ctubbsii I added a new helper method, `modifyPropertiesUnwrapped()` which wraps `modifyProperties()` catches `AccumuloException`, and throws `TableNotFoundException` if thats its cause. I used this new helper method in only the methods that I changed here (setSamplerConfiguration, clearSamplerConfiguration, setLocalityGroups). This adds back the TableNotFoundException that had been removed by a previous change in this PR. It should also restore the exception handling functionality that was in place before converting these methods to use `modifyProperties`.

DomGarguilo committed:
Create new TNFE as to not lose stack info when unwrapping

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1118,6 +1118,27 @@ public Map<String,String> modifyProperties(String tableName,
+   * unwrap and rethrow the TNFE directly. This is a hacky, temporary workaround that we can use
+   * until we are able to change the public API and throw TNFE directly from all applicable methods.
+   */
+  private Map<String,String> modifyPropertiesUnwrapped(String tableName,
+      Consumer<Map<String,String>> mapMutator) throws TableNotFoundException, AccumuloException,
+      AccumuloSecurityException, IllegalArgumentException {
```
Don't need to declare RTEs.

```suggestion
  private Map<String,String> modifyPropertiesUnwrapped(String tableName,
      Consumer<Map<String,String>> mapMutator) throws TableNotFoundException, AccumuloException,
      AccumuloSecurityException {
```

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/clientImpl/TableOperationsImpl.java:
```diff
@@ -1118,6 +1118,27 @@ public Map<String,String> modifyProperties(String tableName,
+    } catch (AccumuloException ae) {
+      Throwable cause = ae.getCause();
+      if (cause instanceof TableNotFoundException) {
+        throw new TableNotFoundException(context.getTableId(tableName).canonical(), tableName,
+            ae.getMessage(), ae);
```
If the table is not found, then looking up the tableId this way is going to throw a completely different TNFE, and the one you are trying to create here will never actually get thrown.

Even if you replace the tableId with null instead of trying to look it up, this construction will mangle the getMessage() on the exception quite a bit, since the parameter here is supposed to be a description snippet that is used to construct a larger message. But what you're passing in here is already a full message from the previous exception.

I think you can just do:
```suggestion
        throw new TableNotFoundException(null, tableName, null, ae);
```

ctubbsii reviewed (changes requested):
The strategy to re-throw a TNFE by unwrapping the exception wrapped in modifyProperties is okay, but the implementation is not working as intended, since you're looking up the table ID. After that is fixed, this could probably be merged.

However, this is still just a hack, because modifyProperties doesn't throw TNFE when it should. A possible better solution is to deprecate it, and create a replacement method, mutateProperties that has the correct exception throwing behavior. So, instead of adding more unwrapping... you remove some of the existing wrapping that is happening in the first place. This would actually go well with the parameter name, which is "mapMutator", which works better with a method named "mutate" rather than "modify".

The problem with creating a new method is that we tend to avoid doing that in bugfix releases, because it violates semver. However, the spirit of semver is to avoid breaking user code. Although this would add a new method, I think that the only code that would break is subclasses, but nobody else is likely to be subclassing the TableOperations interface. So, maybe it is tolerable to add this in a bugfix? If not, your strategy is fine for 2.1 We can fix it the way I'm suggesting in 4.0, and add a deprecation to the old method in a 3.1.

DomGarguilo committed:
Apply suggestions from code review

Co-authored-by: Christopher Tubbs <ctubbsii@apache.org>

DomGarguilo committed:
Formatting

DomGarguilo committed:
Merge remote-tracking branch 'upstream/main' into makeTableOperationsImplAtomic
